### ISSUE TITLE
Make the AI chat not suck

### ISSUE THREAD

**Issue by Visual-Studio-Coder** [zed-industries/zed#18512]
### Check for existing issues

- [X] Completed

### Describe the feature

Its nice that theres markdown and syntax highlighting in the chat window, but I wish it was AS GOOD as GitHub Copilot in the sense that there was a copy button on markdown text, and if there was terminal commands, you could instantly insert it in for example.

Also, the You, Assistant, and System thing makes no sense and isn't intuitive.

**Comment by RustoMCSpit** [comment]
title is funny but please change it to "Adopt Github Co-pilot chat features"

**Comment by Visual-Studio-Coder** [comment]
> title is funny but please change it to "Adopt Github Co-pilot chat features"

Done!

**Comment by JosephTLyons** [comment]
Hey @Visual-Studio-Coder, can you expand on the terminal bullet point a bit - it isn't quite clear to me what you are requesting there.

A copy button on code snippets seems reasonable, especially now since we have the `assistant: copy code` action.

I can't imagine us shifting away from using a plain text editor to a more chat-centric panel. That was something we experimented with, some months back, and just weren't happy with how incredibly inflexible it was. We like that everything is editable and changeable - you can remove parts of the history where the model didn't return anything valuable and regain tokens, etc.

**Comment by Visual-Studio-Coder** [comment]
Oh, that's new information to me. I wasn't aware about the tokens thing, however, its a bit frustrating that it isn't a traditional window. I guess what I'm asking for is a set of actions that make it clear to the user what their editing.

Or you could have a Chat Window vs Editor mode. I find it a bit frustrating that code blocks show up as `````` instead of just a box with monospace font.

For the terminal thing, copilot lets you highlight a codeblock and auto insert it into the terminal

**Comment by Visual-Studio-Coder** [comment]
I have not used GPUI in my life, but I'm going to take a look into making a PR later today. The goal of the PR will be to draw a box around all code blocks and add copy buttons on the code blocks. Don't let this stop you from implementing the same thing because I will likely not be successful at finishing it

**Comment by aaronhma** [comment]
> Hey @Visual-Studio-Coder, can you expand on the terminal bullet point a bit - it isn't quite clear to me what you are requesting there.
> 
> A copy button on code snippets seems reasonable, especially now since we have the `assistant: copy code` action.
> 
> I can't imagine us shifting away from using a plain text editor to a more chat-centric panel. That was something we experimented with, some months back, and just weren't happy with how incredibly inflexible it was. We like that everything is editable and changeable - you can remove parts of the history where the model didn't return anything valuable and regain tokens, etc.

@JosephTLyons, I think a good approach would be how [LM Studio ↗](https://lmstudio.ai/) handles AI responses. By default, the AI's response isn't editable/deleteable, but you can double click or click on a dedicated "Edit message" button. This is available for the both the user and AI's text fields.

<img width="1049" alt="Screenshot 2024-10-02 at 9 16 27 PM" src="https://github.com/user-attachments/assets/e1a426c1-b8f1-406f-9e6c-241ef2eeae86">

**Comment by Visual-Studio-Coder** [comment]
Update: I failed to add the "code container" feature because the `code-folding button` doesn't actually control the AI output. I might try adding a "chat mode" instead.

**Comment by alanxmay** [comment]
I **really** like the way using a plain text editor, it's incredibe.

**Comment by notpeter** [comment]
Code block copy:
- https://github.com/zed-industries/zed/issues/19061
- https://github.com/zed-industries/zed/pull/19312

**Comment by bfulop** [comment]
For me it's exactly the freely editable assistant panel that makes Zed AI so useful and better than anything I've seen. It's pure genius.   
So simple compared to automated, agentic stuff used by Cursor, Windsurf, etc. and works a million times better in practice. 

Please don't remove any of it (including the _Also, the You, Assistant, and System thing makes no sense and isn't intuitive. Please remove that_) 🙏  
I'm using it every day, since the first day this feature came out. It's the killer feature for me.

**Comment by eproxus** [comment]
I agree about the _You_, _Assistant_ and _System_ blocks being a bit confusing. They seem to be buffer "objects" that are editable, but they're not that visible in a wall of text and it is hard to see, for long histories, where each prompt, input and reply starts and stops. I guess this is mostly a UI problem, but I don't know how it could be fixed (color each section until the next block?).

Some actions in the editable buffer doesn't make sense to me, even after reading the documentation again. E.g. I write a question (after the first _You_ object), then get a reply (prefixed by a _Assistant_ object)... then I can change the _You_ to an _Assistant_ object and the _Assistant_ object to a _System_ object by clicking them. When would I need that? Also, if the history is supposed to be fully editable, why can't I insert new blocks?

I get the point about adding text from an _Assistant_ reply into my previous, prompt, but then I'd just rewrite the prompt and remove the _Assistant_ block instead of making it a a _You_ block wholesale (which I think doesn't make sense since it is written as a reply to the question I just wrote).

Also the use of the _System_ block isn't really explained anywhere. What is it for? I would assume it is for system prompts, but why can't I add such a block to prompts in the prompt library then? And what is the difference between a `/prompt` and a _System_ block?

**Comment by andrewhampton** [comment]
I've been a happy Zed user since this summer. I really like the fexibility and integration of the current chat panel.

That said, I have been really impressed with the [Cascade](https://codeium.com/cascade) editor in Windsurf. Specifically, it's able to:

1. Search the current project for relevant context to include
2. Run CLI commands and see the output
3. Be configured with an allow list of commands it can run without approval
4. Create files
5. Edit files
6. Automatically run a plan/execute/evaluate loop involving some/all/none of these capabilities
7. Decide whether or not to use any of these capabilities during the plan/execute/evaluate loop

For example, yesterday, I wrote a script that wrapped rspec and simplecov. Using it would print the rspec test results and the lines not covered by rspec. After adding that script to the allowed list of commands it could run without approval, I could just prompt it with "Please get test coverage to 100% for app/controllers/users_controller.rb" and it would loop on it's own for 5-10 minutes, until the test coverage reached 100%.

This has felt like the first real gear shift in devex I've had since first using copilot in closed beta. I'd love for this sort of functionality to be in Zed, because I much prefer it to VSCode based editors.

**Comment by flitting** [comment]
I love Zed's interface, but I think it would be great if the Markdown responses on Windows could be automatically previewed. It would improve the overall user experience.

**Comment by drdozer** [comment]
The lack of the chat being able to execute commands and directly edit files is why I use vscode + plugin for some things. I really would like to be able to use zed more, but the lack of integration between the chat and the project is an issue. There are various workflows that I would very much like to fully automate or do interactively but can not because of this. Ideally, it would also have access to (some of) the command pallet.

**Comment by hopewise** [comment]
@drdozer I am in a similar situation, where I need Zed to execute commands and be able to feed the command output into context for followup.. is there any plan to implement this great feature ?
